src/runtime/indexes.ts

```diff
@@ -0,0 +1,272 @@
+  aveIndex = createHash();
+  cardinality = 0;
+
+  getOrCreate(parent:any, key:any, createFunc:Function) {
+    let found = parent[key];
+    if(!found) {
+      // if(!ALLOCATION_COUNT["hashlevel"]) ALLOCATION_COUNT["hashlevel"] = 0;
+      // ALLOCATION_COUNT["hashlevel"]++;
+      found = parent[key] = createFunc();
+    }
+    return found;
+  }
+
+  insert(change:Change) {
+    let {getOrCreate} = this;
+    let eIx = getOrCreate(this.eavIndex, change.e, createHash);
+    let aIx = getOrCreate(eIx, change.a, createHash);
+    let vIx = getOrCreate(aIx, change.v, specialCreateArray);
+    vIx.push(change.n, change.transaction, change.round, change.count);
+
+    aIx = getOrCreate(this.aveIndex, change.a, createHash);
+    vIx = getOrCreate(aIx, change.v, createHash);
+    eIx = getOrCreate(vIx, change.e, specialCreateArray);
+    eIx.push(change.n, change.transaction, change.round, change.count);
+
+    this.cardinality++;
+  }
+
+  resolveProposal(proposal:Proposal) {
+    return proposal.info;
+  }
+
+  propose(proposal:Proposal, e:ResolvedValue, a:ResolvedValue, v:ResolvedValue, n:ResolvedValue, transaction = Infinity, round = Infinity) {
+    let forFields = proposal.forFields;
+    forFields.length = 0;
+    if(isResolved(e)) {
+      return this.walkPropose(proposal, this.eavIndex, e, a, v, n, "a", "v", transaction, round);
+    } else if(isResolved(a)) {
+      return this.walkPropose(proposal, this.aveIndex, a, v, e, n, "v", "e", transaction, round);
+    } else {
+      // propose for attribute since that's likely to be the smallest
+      forFields[0] = "a";
+      proposal.info = Object.keys(this.aveIndex);
+      proposal.cardinality = proposal.info.length;
+    }
+    return proposal;
+  }
+
+  walkPropose(proposal:Proposal, index:any, a:ResolvedValue, b:ResolvedValue, c:ResolvedValue, n:ResolvedValue,
+              fieldB:EAVNField, fieldC:EAVNField, transaction:number, round:number):Proposal {
+    let {forFields} = proposal;
+    let bIx = index[a as ID];
+    if(!bIx) {
+      proposal.cardinality = 0;
+      return proposal;
+    }
+    if(isResolved(b)) {
+      let cIx = bIx[b];
+      if(!cIx) {
+        proposal.cardinality = 0;
+        return proposal;
+      }
+      if(isResolved(c)) {
+        let ntrcArray = cIx[c];
+        if(sumTimes(ntrcArray, transaction, round) > 0) {
+          proposal.skip = true;
+          return proposal;
+        } else {
+          proposal.cardinality = 0;
+          return proposal;
+        }
+      } else {
+        forFields[0] = fieldC;
+        proposal.info = Object.keys(cIx);
+        proposal.cardinality = proposal.info.length;
+        return proposal;
+      }
+    } else {
+      forFields[0] = fieldB;
+      proposal.info = Object.keys(bIx);
+      proposal.cardinality = proposal.info.length;
+      return proposal;
+    }
+  }
+
+  walkCheck(index:any, a:ResolvedValue, b:ResolvedValue, c:ResolvedValue, n:ResolvedValue, transaction:number, round:number):boolean {
```

Comment:
This guy deserves a comment along the lines of "check that the deepest resolved level of the index contains at least one entry." Took me a minute to work it out.